### PR TITLE
Add UV cost

### PR DESCRIPTION
**Pull request by luohezhiming**
## Summary/Motivation:

Add costing method for UV system

## Changes proposed in this PR:
- add costing
- add corresponding tests

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/costing/watertap_costing_package.py
Comment (hunterbarber) [inline watertap/costing/watertap_costing_package.py]: Not sure `uv_aop_stack` is the nomenclature that best represents the UV reactor configuration. ED is costed as such because it seems to be completely modular, so the stack of cells makes sense. Here, something like `uv_aop_lamp_bundle` would be a physical equivalent, but the model is not being costed as such and therefore I'm not sure of the reason this is presented as its own function, similar to the ED costing. Maybe you and others can comment for clarification.

File: watertap/costing/watertap_costing_package.py
Comment (hunterbarber) [inline watertap/costing/watertap_costing_package.py:1385]: Little confused with how `blk.electricity_cost * electricity_demand` is a capital cost, just thinking that some electricity demand in kW (or GJ/yr or any equivalent) should be a fixed operating cost. Are you costing the uv (lamp?) replacement as a function of the electricity demand? If so, why is the equation not something similar to:
`blk.fixed_operating_cost_constraint = pyo.Constraint(
        expr=blk.fixed_operating_cost
        == blk.factor_uv_replacement * blk.uv_replacement_cost * electricity_demand + blk.electricity_cost * electricity_demand
    )`

File: watertap/unit_models/tests/test_uv_aop.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_uv_aop.py]: At 1 kg/s of feed flow, is this reasonable cost and is there any need to also test an industrial scale?

**Review by hunterbarber — COMMENTED** [review]
General comments on costing implementation provided

File: watertap/costing/watertap_costing_package.py
Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: I've changed to `cost_uv_aop_bundle` as suggested, the UV system include uv reactor, lamps and other equipment. I previously thought "stack" has a similar meaning with "bundle", but I agree "bundle" is more appropriate in this case.

File: watertap/costing/watertap_costing_package.py
Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py:1385]: Electricity is registered as another operating cost flow in the `watertap_costing_package`, and it does not belong to the fixed operating cost (should be "annual operating cost" or something else). And you are correct, the `blk.electricity_cost * electricity_demand` represents lamps and other relevant equipment capital cost which is sized by electricity demand, and the fixed operating costs are simply consumable cost of these devices.

File: watertap/unit_models/tests/test_uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/tests/test_uv_aop.py]: I've changed the value to account for an industrial scale.

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [inline watertap/costing/watertap_costing_package.py]:
```suggestion
def cost_uv_aop_bundle(blk, flow_cost, lamp_cost, factor_uv_replacement):

```

File: watertap/costing/watertap_costing_package.py
Comment (hunterbarber) [inline watertap/costing/watertap_costing_package.py]: If you look across the __init__, readthedocs, and `from watertap.unit_models import` sequences of unit models, UV-AOP appears in a different location each time. There isn't a standardization in the order these appear I think.

File: watertap/costing/watertap_costing_package.py
Comment (hunterbarber) [inline watertap/costing/watertap_costing_package.py]: Is this actually still TODO?

File: watertap/costing/watertap_costing_package.py
Comment (hunterbarber) [inline watertap/costing/watertap_costing_package.py]: Here again, seems like the order models go: copied from IDAES models, then WaterTAP unique models in the order they appear in the __init__ and readthedocs. Obviously minor but will be more relevant as this list becomes longer.

File: watertap/unit_models/uv_aop.py
Comment (hunterbarber) [inline watertap/unit_models/uv_aop.py]: What are practical values for the number of reactors? Would it always be between 1 and ~10? Not going to affect the convergence too much, since the scaled variable would then be between 100 and ~1000 respectively but can you comment on the reasoning for this sf?

**Review by hunterbarber — APPROVED** [review]
Left a couple comments but all were extremely minor and may not even need changed therefore I'm marking this as approved. May be good to get some discussion on the couple of comments before merging though.

File: watertap/costing/watertap_costing_package.py
Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: Typically, we arrange the sequence alphabetically, but since there are not so many detailed models, so people just add them following the last one. I add it at the first just for avoiding potential merging conflicts because I believe there are other detailed model PRs (e.g. yours and Kurby's). If we all add it at the last, then we will have merging conflicts. But I agree with you that we can follow the sequence in the **init** file at present.

File: watertap/unit_models/uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/uv_aop.py]: Typically in a UV system, there are 1-5 reactors. So the sf should be 1. I copied from my last sentence and forgot to change the value.

File: watertap/costing/watertap_costing_package.py
Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: Deleted

File: watertap/costing/watertap_costing_package.py
Comment (hunterbarber) [reply watertap/costing/watertap_costing_package.py]: @TimBartholomew @adam-a-a Is it worth creating an issue for standardizing the order of the unit models when presented as a list in various locations? See `watertap/docs/technical_reference/unit_models/index.rst`, `watertap/watertap/unit_models/__init__.py`, and `watertap/watertap/costing/watertap_costing_package.py -> imports` as examples that all appear differently. This is extremely minor and more ease of use I would think, but may be worth converting to alphabetical as the list gets longer and as @luohezhiming said, to more consistently avoid merge conflicts with new models' PRs. Currently it seemed as though the list had the newest models at the end of the list but that has gotten twisted as the aforementioned issues occur.

Comment (bknueven) [reply watertap/costing/watertap_costing_package.py]: We should generally do things alphabetically. This, along with the way `black` formats the code, will very often avoid merge conflicts.

File: watertap/unit_models/uv_aop.py
Comment (bknueven) [inline watertap/unit_models/uv_aop.py]: This whole set-up doesn't make a huge amount of sense to me from a mathematical modeling perspective. What quantity should `electricity_demand` be equal to?

If it's the `max(b.electricity_demand_phase_comp[t, "Liq", :])` then we need to be doing something different.

File: watertap/unit_models/uv_aop.py
Comment (bknueven) [inline watertap/unit_models/uv_aop.py]:
```suggestion
```

**Review by bknueven — CHANGES_REQUESTED** [review]
I need some clarification on the model.

File: watertap/unit_models/uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/uv_aop.py]: The `electricity_demand` is equal to `max(b.electricity_demand_phase_comp[t, "Liq", :])`. All of these set-up is to avoid using `max` function because it is not continuous and may have problems when using IDAES ipopt solver. So I follow the way in this https://github.com/IDAES/idaes-pse/blob/d1a2add96bf660681048b05ea8e90fb4b51294ee/idaes/models/unit_models/mixer.py#L632

Comment (bknueven) [reply watertap/unit_models/uv_aop.py]: This implementation is a bit different than the one you point to in IDAES, which looks correct to me.

I would also change the variable name (from min to max) to be less confusing.

I can think of another implementation which might be a bit more efficient, if you have a lot of values. @andrewlee94 I wonder if we should add a helper in IDAES to deal with these recursive min/max formulations, since we have at least two examples.

Comment (andrewlee94) [reply watertap/unit_models/uv_aop.py]: We could add a helper function for this, although I am always a bit concerned about making things like this too easy to use for users who don't know the issues associated with them. PRs are always welcome.

On a different topic, I note that all the doc strings refer to a minimum energy demand whilst the function itself is a max. Whilst this does make some sense in the global process level if you stop and think about it for a while, in the more local sense you are really finding the maximum of the potential electricity demands. I think it might make more sense to call this `max_component_electricity_demand` or something similar to better indicate what this represents.

Comment (bknueven) [reply watertap/unit_models/uv_aop.py]: Your second comment brings up a good point -- if this is really the "minimum" energy the process could use, but it could use more, then we could use a standard inequality representation and skip using `smooth_max`.

Comment (andrewlee94) [reply watertap/unit_models/uv_aop.py]: @bknueven I am not sure that is a good idea - it leaves the energy unbounded to the upper end without any additional constraints. My initial though as I was reading this was that this would give us the minimum theoretical energy required, and the unit model could then add an efficiency onto this to get the actual energy demand - this would require that this be well-defined and not an inequality.

Comment (luohezhiming) [reply watertap/unit_models/uv_aop.py]: I agree that we should change the name to `max_component_electricity_demand` because bas an intermedium vairable it is easier for understanding. @bknueven we talked about the inequality representation before, but we think it might have bounding issues when doing optimization problems.

File: watertap/unit_models/uv_aop.py
Comment (bknueven) [inline watertap/unit_models/uv_aop.py]:
```suggestion
                        b.max_phase_electricity_demand[
```

File: watertap/unit_models/uv_aop.py
Comment (bknueven) [inline watertap/unit_models/uv_aop.py]:
```suggestion
                        b.max_component_electricity_demand[
```

**Review by bknueven — DISMISSED** [review]
My original comment was that the formulation which calculates the `max` is incorrect, which wasn't caught because it is not tested, and even with a test may not have been caught.

Hence my suggestion for a utility, which we could at least test/verify separately, and then use where needed. This would be especially useful if this is a modeling paradigm we need to repeat elsewhere. However, it would be enough to just correct and test it in this unit model.

**Comment by luohezhiming** [comment]
> My original comment was that the formulation which calculates the `max` is incorrect, which wasn't caught because it is not tested, and even with a test may not have been caught.
> 
> Hence my suggestion for a utility, which we could at least test/verify separately, and then use where needed. This would be especially useful if this is a modeling paradigm we need to repeat elsewhere. However, it would be enough to just correct and test it in this unit model.

Thank you for catching this mistake that I did realize before! I've committed your suggested change.

**Comment by luohezhiming** [comment]
@bknueven The Checks for macOS setup always directly cancelled, I see your PR #635 also cancelled this check, do you know what happened?

**Comment by hunterbarber** [comment]
> @bknueven The Checks for macOS setup always directly cancelled, I see your PR #635 also cancelled this check, do you know what happened?

Looks like the following is happening with [macOS tests outlined here with scheduled brownouts](https://github.com/actions/virtual-environments/issues/5583). My PR #617 is now also encountering this issue when updating the branch although it is ready to merge and was previously passing all tests. @lbianchi-lbl Does any action need taken with this?

**Comment by lbianchi-lbl** [comment]
> > @bknueven The Checks for macOS setup always directly cancelled, I see your PR #635 also cancelled this check, do you know what happened?
> 
> Looks like the following is happening with [macOS tests outlined here with scheduled brownouts](https://github.com/actions/virtual-environments/issues/5583). My PR #617 is now also encountering this issue when updating the branch although it is ready to merge and was previously passing all tests. @lbianchi-lbl Does any action need taken with this?

@hunterbarber Thanks a lot for the heads-up, I remember seeing some kind of deprecation warnings but I had no idea that the recent cancellations were due to deliberate brown-outs. I'll open a PR to update the macOS runner version.

**Comment by adam-a-a** [comment]
> > My original comment was that the formulation which calculates the `max` is incorrect, which wasn't caught because it is not tested, and even with a test may not have been caught.
> > Hence my suggestion for a utility, which we could at least test/verify separately, and then use where needed. This would be especially useful if this is a modeling paradigm we need to repeat elsewhere. However, it would be enough to just correct and test it in this unit model.
> 
> Thank you for catching this mistake that I did realize before! I've committed your suggested change.

Actually, correct me if I'm wrong, but I think @bknueven was referring to his suggested changes on one of the arguments in the smooth_max function, which I don't think @luohezhiming incorporated. However, from just a glance and recalliing the original intention, I'm not sure those changes need to be made because after using smooth_max, later on, the last element which contains that actual max is grabbed and assigned to the actual energy consumption. 

Perhaps with @bknueven 's suggestion, we would not need to grab the last element -- not sure without trying myself -- but it looks that way.

**Comment by adam-a-a** [comment]
> > > My original comment was that the formulation which calculates the `max` is incorrect, which wasn't caught because it is not tested, and even with a test may not have been caught.
> > > Hence my suggestion for a utility, which we could at least test/verify separately, and then use where needed. This would be especially useful if this is a modeling paradigm we need to repeat elsewhere. However, it would be enough to just correct and test it in this unit model.
> > 
> > 
> > Thank you for catching this mistake that I did realize before! I've committed your suggested change.
> 
> Actually, correct me if I'm wrong, but I think @bknueven was referring to his suggested changes on one of the arguments in the smooth_max function, which I don't think @luohezhiming incorporated. However, from just a glance and recalliing the original intention, I'm not sure those changes need to be made because after using smooth_max, later on, the last element which contains that actual max is grabbed and assigned to the actual energy consumption.
> 
> Perhaps with @bknueven 's suggestion, we would not need to grab the last element -- not sure without trying myself -- but it looks that way.

Nevermind-- I see now that it wouldn't have worked out, and although it appeared that @bknueven 's suggestions weren't adopted by looking at the conversation, that was outdated.

**Comment by bknueven** [comment]
I won't hold this up anymore, but I am still concerned about this max formulation not being covered by tests.

If we merge this as-is, we should open an issue to track the uncovered code.

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py:726]: @luohezhiming Based on @bknueven 's previous suggestion, I think it'd be good to have a quick test checking that max energy consumption gets selected as intended, so maybe a separate test with two target components (two EE/Os), and assert that the expected energy consumption is set to the actual energy consumption.

Comment (adam-a-a) [reply watertap/unit_models/tests/test_uv_aop.py:726]: Not sure if you did include that somewhere and I missed it...if not, do you think you can add that?

Comment (luohezhiming) [reply watertap/unit_models/tests/test_uv_aop.py:726]: I could try to add another dummy component to the test to validate the test.

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]:
```suggestion
```

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [inline watertap/costing/watertap_costing_package.py]: Is this representative of UV _lamp_ replacement?

Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: I sum up the consumable costs of lamps, sleeves, ballasts and sensors, so it accounts for all of them.

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [inline watertap/costing/watertap_costing_package.py]: Can we rename this `factor_uv_lamp_replacement` so that it is immediately obvious?

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]: Which unused var is it that gets used for costing? I only ask because the first thought that comes to mind is that maybe that var should only be in the costing package but not sure.

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]: I don't have a problem with this line being here, but we typically delete these print lines since they are meant for debugging purposes and we usually clean this kind of stuff up before merging.

File: watertap/unit_models/tests/test_uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/tests/test_uv_aop.py]: The unused variable is `num_of_reactors`, since the costing data I used is for a uv system, so I add this `num_of_reactors` to account for number of reactors in a uv system.

File: watertap/costing/watertap_costing_package.py
Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: Yeah, I think it is a good idea to be more clear

File: watertap/unit_models/tests/test_uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/tests/test_uv_aop.py]: Deleted

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]: You could replace this line with `assert_optimal_termination(results)`, which I think checks that `TerminationCondition.optimal` is true _and_ that solver status is OK

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]: same as above- basically, replace with assert_optimal_termination or add assertion to check that solver status is OK

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]: same comment as before regarding print lines

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py:723]: Great!

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_uv_aop.py]: same comment as before regarding solver status OK

File: watertap/unit_models/uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/uv_aop.py]: Can you change all these small lower bounds (1e-18) to 0 and ensure solver convergence is still achieved?

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [reply watertap/costing/watertap_costing_package.py]:
```suggestion
            doc="UV replacement factor accounting for lamps, sleeves, ballasts and sensors [fraction of uv replaced/year]",

```

File: watertap/unit_models/tests/test_uv_aop.py
Comment (adam-a-a) [reply watertap/unit_models/tests/test_uv_aop.py]: Does this represent number of reactors in series, parallel, or both? If parallel, would the var end up being used in the model without costing?

File: watertap/unit_models/uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/uv_aop.py]: Yeah, I've revised all these bounds to `(0,None) `and all tests are passed.

File: watertap/unit_models/tests/test_uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/tests/test_uv_aop.py]: It represent reactors in parallel. This var is onlly used for costing the uv reactor.

Comment (luohezhiming) [reply watertap/unit_models/tests/test_uv_aop.py]: I think we could keep this `num_of_reactors` as a unit model variable because we could use it for the uv dose expression (related to UVT, flow rate and number of banks).

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [inline watertap/costing/watertap_costing_package.py]: I'm a little unsure about this. This seems like the control volume for your UV model only accounts for 1 reactor, so you multiply by number of (parallel) reactors to accommodate a system with higher capacity.... but then the lamp_cost is multiplied by power demand, which I would expect to increase as well with number of reactors. Moreover, couldn't our control volume represent the whole system of reactors without having to ever specify number of reactors?

Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: This is a pretty good point! Actually different uv reactors have different numbers of lamps and other equipments, but as Tim suggested, I'd better not add so many variablles such as number of lamps, etc. I think for a simple solution, we could delete this `num_of_reactors` variable and use our control volume to represent the whole uv system as you suggested.

File: watertap/unit_models/uv_aop.py
Comment (adam-a-a) [inline watertap/unit_models/uv_aop.py:541]: I suppose we are missing a test to check that the solution works as intended with more than one phase. Nevertheless, I think this was already an additional capability that you didn't need to add, so maybe we can just make an issue to keep track of any UV-related things we want to revisit in future PRs for later improvement. In this case, we can note that a test should be added to check electricity demand for a multi-component and multi-phase example.

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [reply watertap/costing/watertap_costing_package.py]: OK, then it seems this is the only further change needed before merging, considering that you already addressed my other comments.

**Review by adam-a-a — CHANGES_REQUESTED** [review]
Excellent job on this. I think we should be ready to merge this once we remove the `num_of_reactors` being used in the capital cost constraint. Please also note things that we might want to consider in future PRs by adding to an issue for UV. We can have one master issue for all things related to UV.

File: watertap/costing/watertap_costing_package.py
Comment (adam-a-a) [reply watertap/costing/watertap_costing_package.py]: _Unless_ the number of reactors accounts for redundant reactors that aren't used all at the same time. Then I could understand why only 1 reactor is considered for energy demand, while 3 (in your test examples) are considered for capital cost, which could represent 1 operational reactor and 2 redundant.

File: watertap/unit_models/uv_aop.py
Comment (luohezhiming) [reply watertap/unit_models/uv_aop.py:541]: I don't know what kind of waterborne pathogens is solid or gas phase, maybe we could add a dummy solute for verifying the test? e.g. solid `DCE`

Comment (adam-a-a) [reply watertap/unit_models/uv_aop.py:541]: We could try something like that. Another option could be to use the IDAES generic property framework to define liquid and solid. Whichever path we take, I think this test can be reserved for a subsequent PR and noted in your master issue for UV, but if you think it'd be quick and easy for you to implement, I won't stop you!

File: watertap/costing/watertap_costing_package.py
Comment (luohezhiming) [reply watertap/costing/watertap_costing_package.py]: @adam-a-a in the given example, there are 3 parallel reactors and 1 is redundant, there are 15 average number of lamps working over the simulation. Overall, the energy demand depends on the working lamps but the capital cost includes the redundant reactor, lamps and other equipments.

**Comment by luohezhiming** [comment]
> Excellent job on this. I think we should be ready to merge this once we remove the `num_of_reactors` being used in the capital cost constraint. Please also note things that we might want to consider in future PRs by adding to an issue for UV. We can have one master issue for all things related to UV.

@adam-a-a I've deleted the `num_of_reactors` var and addressed all other comments. I will create a master issue for further refinement.

**Review by adam-a-a — APPROVED** [review]
Nice job!

**Pull request merged**
